network-libp2p/src/connection_pool/behaviour.rs

```diff
@@ -115,6 +118,32 @@ struct ConnectionState<T> {
     required_services: Services,
 }
 
+enum ConnectionStateAction {
+    UnbanPeer(PeerId),
+}
+
+impl Stream for ConnectionState<PeerId> {
+    type Item = ConnectionStateAction;
+
+    fn poll_next(self: Pin<&mut Self>, _: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.banned.is_empty() {
+            return Poll::Pending;
+        }
+
+        if let Some((peer_id, _)) = self.banned.iter().find(|(_, banned_since)| {
+            if banned_since.elapsed() >= self.ban_time {
+                return true;
+            }
+
+            false
+        }) {
+            return Poll::Ready(Some(ConnectionStateAction::UnbanPeer(*peer_id)));
+        }
+
+        Poll::Pending
+    }
+}
```

Comment:
(Already talked to @Eligioo about this:) The `poll_next` function doesn't register with a waker, so in isolation, it wouldn't be called again.